vfio_user/src/lib.rs

```diff
@@ -782,3 +791,482 @@ impl Client {
             .map_err(Error::StreamShutdown)
     }
 }
+
+pub trait ServerBackend {
+    fn region_read(
+        &mut self,
+        _region: u32,
+        _offset: u64,
+        _data: &mut [u8],
+    ) -> Result<(), std::io::Error>;
+    fn region_write(
+        &mut self,
+        _region: u32,
+        _offset: u64,
+        _data: &[u8],
+    ) -> Result<(), std::io::Error>;
+    fn dma_map(
+        &mut self,
+        _flags: DmaMapFlags,
+        _offset: u64,
+        _address: u64,
+        _size: u64,
+        _fd: Option<&File>,
+    ) -> Result<(), std::io::Error>;
+    fn dma_unmap(
+        &mut self,
+        _flags: DmaUnmapFlags,
+        _address: u64,
+        _size: u64,
+    ) -> Result<(), std::io::Error>;
+    fn reset(&mut self) -> Result<(), std::io::Error>;
+    fn set_irqs(
+        &mut self,
+        _index: u32,
+        _flags: u32,
+        _start: u32,
+        _count: u32,
+        _fds: Vec<File>,
+    ) -> Result<(), std::io::Error>;
+}
+
+pub struct Server {
+    listener: UnixListener,
+    resettable: bool,
+    irqs: Vec<IrqInfo>,
+    regions: Vec<vfio_region_info>,
+}
+
+impl Server {
+    pub fn new(
+        path: &Path,
+        resettable: bool,
+        irqs: Vec<IrqInfo>,
+        regions: Vec<vfio_region_info>,
+    ) -> Result<Server, Error> {
+        let listener = UnixListener::bind(path).map_err(Error::SocketBind)?;
+
+        Ok(Server {
+            listener,
+            resettable,
+            irqs,
+            regions,
+        })
+    }
+
+    pub fn run(&self, backend: &mut dyn ServerBackend) -> Result<(), Error> {
+        let (mut stream, _) = self.listener.accept().map_err(Error::SocketAccept)?;
+
+        loop {
+            let mut header = Header::default();
+
+            let mut fds = vec![0; 16];
+            let mut iovecs = vec![iovec {
+                iov_base: header.as_mut_slice().as_mut_ptr() as *mut c_void,
+                iov_len: header.as_mut_slice().len(),
+            }];
+            // SAFETY: Safe as the iovect is correctly initialised and fds is big enough
+            let (bytes, fds_received) = unsafe {
+                stream
+                    .recv_with_fds(&mut iovecs, &mut fds)
+                    .map_err(Error::ReceiveWithFd)?
+            };
+
+            // Other end closed connection
+            if bytes == 0 {
+                info!("Connection closed");
+                break;
+            }
+
+            fds.resize(fds_received, 0);
+
+            let fds: Vec<File> = fds
+                .iter()
+                // SAFETY: Safe as we have only valid FDs in the vector now
+                .map(|fd| unsafe { File::from_raw_fd(*fd) })
+                .collect();
+
+            match header.command {
```

Comment:
I think we should always send an `Error` reply to the client if the server stops running due to errors. One idea is we can put this big `match` block into a `catch_unwind` statement, form which we always catch any errors and inform the client before stop running the server. Also, it would simply the `match` block by avoiding sending "Error" replies individually from some of the commands.

Reply:
Great suggestion!